### ISSUE TITLE
afterChange invoked with 2d changes then single observe change

### ISSUE THREAD

**Issue by aaronbeall** [handsontable/handsontable#5011]
### Description

According to [the docs](https://docs.handsontable.com/0.16.1/Hooks.html#event:afterChange) `afterChange` should be invoked with a "2D array containing information about each of the edited cells", however after enabling `columnSorting` I found that editing the table results in *two* callbacks:

1. The first `afterChange` is called with a 2d array of changes and source `"edit"` as expected.
2. The second `afterChange` is called with a single change (4 values describing the cell edit) and source `"ObserveChanges.change"`. This led to bugs in how I handle `afterChange`. 

### Steps to reproduce

1. Render `react-handsontable` like this:
```
        // `data` is an array of objects
        <HotTable root="hot"
                      data={ data }
                      colHeaders={ colHeaders }
                      columns={ columns }
                      afterChange={ this.handleAfterChange }
                      columnSorting={ true }
                      sortIndicator={ true }
        />
``` 
2. Make a change to the cell at (0,0)
3. `handleAfterChange` is invoked twice, once with 2d array of changes, then with single change.

The second `afterChange` is only invoked if `columnSorting={true}` and `colHeaders` is set to an array of HTML strings with row count in them (it doesn't happen if `colHeaders={true}`).

I know that `columnSorting` internally turns on `observeChanges` but the behavior of `afterChange` to invoke with a single change and not a 2d array is unexpected.

### Demo

https://jsfiddle.net/handsoncode/3gas9z6q/

### Your environment
* Handsontable version: `react-handsontable` 0.3.1 and `handsontable` 0.33.0
* Browser Name and version: Chrome 65
* Operating System: MacOS High Sierra

**Comment by aaronbeall** [comment]
@AMBudnik I wasn't able to get it to happen in the demo. I've stripped away a lot of the other things happening in my usage, but what I can tell you is:
- Yes, I'm using the React wrapper with the render code I posted above.
- My `afterChange` handler updates some data which causes the `colHeaders` to change (replaces the array with a new array with different strings). This causes `HotTable` to re-render, which I've wrapped in a pure component so it doesn't re-render unless one of the props has actually changed.
- If I remove the `columnSorting={true}` the second "ObserveChanges" never fires.
- If I remove the update to `colHeaders` from the `afterChange` handler the second "ObserveChanges" never fires. 

Is there a way I could debug Handsontable to figure out *what change* exactly is causing the "ObserveChanges" to fire?

**Comment by aaronbeall** [comment]
@AMBudnik I was able to reproduce the behavior on codepen based on the `react-handsontable` demo:

https://codepen.io/abeall/pen/eMxbzQ

The second `"ObserveChanges"` fires if you edit a field. What else I noticed is that in `0.33.0` I get the issue that the `ObserveChanges.change` is not a 2d array:

![img](http://g.recordit.co/5AsXitRmpZ.gif)

But if I change the version to the latest `0.38.1` I get a 2d array in the second "ObserveChanges" event:

![img](http://g.recordit.co/0LtDCvGCZx.gif)

So, I guess it's solved? The behavior I'm seeing even with `0.38.1` isn't ideal but I guess it makes sense, I will need to filter out all `"ObserveChanges"` to avoid re-acting to my own changes...

**Comment by AMBudnik** [comment]
Yes, it does. We need to have one call for cell edition and then another if you call `updateColHeaders` inside `afterChange` hook, as currently for a single change a whole table is being rerendered.
